sim/midas/src/main/scala/midas/core/CPUManagedStreamEngine.scala

```diff
@@ -0,0 +1,213 @@
+
+// See LICENSE for license details.
+
+package midas.core
+
+import chisel3._
+import chisel3.util._
+import chisel3.experimental.IO
+import freechips.rocketchip.amba.axi4._
+import freechips.rocketchip.config.{Parameters, Field}
+import freechips.rocketchip.diplomacy._
+import freechips.rocketchip.util.DecoupledHelper
+
+import midas.widgets._
+import midas.widgets.CppGenerationUtils._
+
+/**
+  * A helper container to serialize per-stream constants to the header. This is
+  * currently somewhat redundant with the default header emission for widgets.
+  */
+case class StreamDriverParameters(
+  channelName: String,
+  bufferBaseAddress: Int,
+  mmioRegisters: (String, Int)*) {
+
+  def headerFragment(mmioBaseAddress: BigInt): String = {
+    val registerMacros = for ((name, offset) <- mmioRegisters) yield {
+      genConstStatic(s"${channelName}_${name}_address", UInt32(mmioBaseAddress + offset))
+    }
+    val dmaAddressMacro = genConstStatic(s"${channelName}_dma_address", UInt32(bufferBaseAddress))
+
+    (dmaAddressMacro +: registerMacros.toSeq).mkString
+  }
+}
+
+class CPUManagedStreamEngine(p: Parameters, val params: StreamEngineParameters) extends StreamEngine(p) {
+
+  val dmaBytes = p(DMANastiKey).dataBits / 8
+  val pcisNodeOpt = Some(AXI4SlaveNode(
+      Seq(AXI4SlavePortParameters(
+        slaves = Seq(AXI4SlaveParameters(
+          address       = Seq(AddressSet(0, (BigInt(1) << p(DMANastiKey).dataBits) - 1)),
```

Comment:
I'm fine punting it to another PR to help merge this in. For my understanding, `DMANastiKey` should be purged because it just sets the parameters for the CPU-managed DMA (AXI) - which is all done in this file? Would this be used somewhere else or would this just be here? Shouldn't there be different "DMANasti" parameters for potential different DMA cases or would a new stream engine be made?

Reply:
By purge i mean, replace with a new `Field` as i have done for the DRAM-subsystem that doesn't use `Nasti` classes, but a dedicated class, say `CPUManagedAXI4Params` or `PCISParams` or something to this effect, that captures all of the true configuration options for the interface. Also better names.